// translation.proto
// Baketa Translation Service gRPC Interface Definition
// Version: 1.0.0
// Phase 2.1: Protoãƒ•ã‚¡ã‚¤ãƒ«è¨­è¨ˆ

syntax = "proto3";

package baketa.translation.v1;

import "google/protobuf/timestamp.proto";

// ============================================================================
// Enums
// ============================================================================

/// <summary>
/// ç¿»è¨³ã‚¨ãƒ©ãƒ¼ã®ç¨®é¡
/// </summary>
enum TranslationErrorType {
  TRANSLATION_ERROR_TYPE_UNSPECIFIED = 0;  // Proto3 requires zero value
  TRANSLATION_ERROR_TYPE_UNKNOWN = 1;
  TRANSLATION_ERROR_TYPE_NETWORK = 2;
  TRANSLATION_ERROR_TYPE_AUTHENTICATION = 3;
  TRANSLATION_ERROR_TYPE_QUOTA_EXCEEDED = 4;
  TRANSLATION_ERROR_TYPE_ENGINE = 5;
  TRANSLATION_ERROR_TYPE_UNSUPPORTED_LANGUAGE = 6;
  TRANSLATION_ERROR_TYPE_INVALID_INPUT = 7;
  TRANSLATION_ERROR_TYPE_TIMEOUT = 8;
  TRANSLATION_ERROR_TYPE_EXCEPTION = 9;
  TRANSLATION_ERROR_TYPE_NETWORK_ERROR = 10;
  TRANSLATION_ERROR_TYPE_PROCESSING_ERROR = 11;
  TRANSLATION_ERROR_TYPE_SERVICE_UNAVAILABLE = 12;
  TRANSLATION_ERROR_TYPE_MODEL_LOAD_ERROR = 13;
  TRANSLATION_ERROR_TYPE_INVALID_REQUEST = 14;
  TRANSLATION_ERROR_TYPE_INVALID_RESPONSE = 15;
  TRANSLATION_ERROR_TYPE_OPERATION_CANCELED = 16;
  TRANSLATION_ERROR_TYPE_UNEXPECTED_ERROR = 17;
  TRANSLATION_ERROR_TYPE_AUTH_ERROR = 18;
  TRANSLATION_ERROR_TYPE_RATE_LIMIT_EXCEEDED = 19;
  TRANSLATION_ERROR_TYPE_MODEL_ERROR = 20;
}

// ============================================================================
// Basic Types
// ============================================================================

/// <summary>
/// çŸ©å½¢é ˜åŸŸã‚’è¡¨ã™
/// </summary>
message Rectangle {
  int32 x = 1;
  int32 y = 2;
  int32 width = 3;
  int32 height = 4;
}

/// <summary>
/// ç¿»è¨³ã§ä½¿ç”¨ã™ã‚‹è¨€èªæƒ…å ±
/// </summary>
message Language {
  string code = 1;           // è¨€èªã‚³ãƒ¼ãƒ‰ (ä¾‹: "en", "ja", "zh-CN")
  string display_name = 2;   // è¡¨ç¤ºå (ä¾‹: "English", "æ—¥æœ¬èª")
  string name = 3;           // è¨€èªåï¼ˆè‹±èªï¼‰
  string native_name = 4;    // è¨€èªåï¼ˆç¾åœ°èªï¼‰
  string region_code = 5;    // åœ°åŸŸã‚³ãƒ¼ãƒ‰ (ä¾‹: "US", "JP")
  bool is_auto_detect = 6;   // è‡ªå‹•æ¤œå‡ºã‹ã©ã†ã‹
  bool is_right_to_left = 7; // RTLè¨€èªã‹ã©ã†ã‹
}

// ============================================================================
// Complex Types
// ============================================================================

/// <summary>
/// ç¿»è¨³ã‚³ãƒ³ãƒ†ã‚­ã‚¹ãƒˆ
/// </summary>
message TranslationContext {
  string game_profile_id = 1;          // ã‚²ãƒ¼ãƒ ãƒ—ãƒ­ãƒ•ã‚¡ã‚¤ãƒ«ID
  string scene_id = 2;                 // ã‚·ãƒ¼ãƒ³è­˜åˆ¥å­
  string dialogue_id = 3;              // ä¼šè©±ID
  Rectangle screen_region = 4;         // ç”»é¢é ˜åŸŸ
  repeated string tags = 5;            // ã‚³ãƒ³ãƒ†ã‚­ã‚¹ãƒˆã‚¿ã‚°
  int32 priority = 6;                  // ã‚³ãƒ³ãƒ†ã‚­ã‚¹ãƒˆå„ªå…ˆåº¦ï¼ˆ0ï½100ï¼‰
  map<string, string> additional_context = 7; // è¿½åŠ ã‚³ãƒ³ãƒ†ã‚­ã‚¹ãƒˆæƒ…å ±ï¼ˆJSONæ–‡å­—åˆ—åŒ–ï¼‰
  string genre = 8;                    // ãƒ†ã‚­ã‚¹ãƒˆã®ã‚¸ãƒ£ãƒ³ãƒ«
  string domain = 9;                   // ãƒ†ã‚­ã‚¹ãƒˆã®ãƒ‰ãƒ¡ã‚¤ãƒ³
}

/// <summary>
/// ç¿»è¨³ã‚¨ãƒ©ãƒ¼æƒ…å ±
/// </summary>
message TranslationError {
  string error_code = 1;               // ã‚¨ãƒ©ãƒ¼ã‚³ãƒ¼ãƒ‰
  string message = 2;                  // ã‚¨ãƒ©ãƒ¼ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸
  string details = 3;                  // è©³ç´°æƒ…å ±
  bool is_retryable = 4;               // å†è©¦è¡Œå¯èƒ½ã‹ã©ã†ã‹
  TranslationErrorType error_type = 5; // ã‚¨ãƒ©ãƒ¼ã®ç¨®é¡

  // Exceptionæƒ…å ±ï¼ˆC# Exceptionå‹ã¯ç›´æ¥ã‚·ãƒªã‚¢ãƒ©ã‚¤ã‚ºä¸å¯ï¼‰
  string exception_type = 6;           // ä¾‹å¤–ã®å‹å
  string exception_message = 7;        // ä¾‹å¤–ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸
  string exception_stack_trace = 8;    // ã‚¹ã‚¿ãƒƒã‚¯ãƒˆãƒ¬ãƒ¼ã‚¹
}

// ============================================================================
// Request/Response Messages - Single Translation
// ============================================================================

/// <summary>
/// å˜ä¸€ç¿»è¨³ãƒªã‚¯ã‚¨ã‚¹ãƒˆ
/// </summary>
message TranslateRequest {
  string source_text = 1;              // ç¿»è¨³å…ƒãƒ†ã‚­ã‚¹ãƒˆ (æœ€é »å‡ºã€1ãƒã‚¤ãƒˆã‚¨ãƒ³ã‚³ãƒ¼ãƒ‰)
  Language source_language = 2;        // å…ƒè¨€èª
  Language target_language = 3;        // å¯¾è±¡è¨€èª
  string request_id = 4;               // ãƒªã‚¯ã‚¨ã‚¹ãƒˆID (UUIDæ–‡å­—åˆ—)
  TranslationContext context = 5;      // ç¿»è¨³ã‚³ãƒ³ãƒ†ã‚­ã‚¹ãƒˆï¼ˆoptionalï¼‰
  string preferred_engine = 6;         // å„ªå…ˆã™ã‚‹ç¿»è¨³ã‚¨ãƒ³ã‚¸ãƒ³åï¼ˆoptionalï¼‰
  map<string, string> options = 15;    // ãƒªã‚¯ã‚¨ã‚¹ãƒˆã‚ªãƒ—ã‚·ãƒ§ãƒ³ï¼ˆJSONæ–‡å­—åˆ—åŒ–ï¼‰
  google.protobuf.Timestamp timestamp = 16; // ãƒªã‚¯ã‚¨ã‚¹ãƒˆã‚¿ã‚¤ãƒ ã‚¹ã‚¿ãƒ³ãƒ—
}

/// <summary>
/// å˜ä¸€ç¿»è¨³ãƒ¬ã‚¹ãƒãƒ³ã‚¹
/// </summary>
message TranslateResponse {
  string request_id = 1;               // å¯¾å¿œã™ã‚‹ãƒªã‚¯ã‚¨ã‚¹ãƒˆID (UUIDæ–‡å­—åˆ—)
  string source_text = 2;              // ç¿»è¨³å…ƒãƒ†ã‚­ã‚¹ãƒˆ
  string translated_text = 3;          // ç¿»è¨³çµæœãƒ†ã‚­ã‚¹ãƒˆ
  Language source_language = 4;        // ç¿»è¨³å…ƒè¨€èª
  Language target_language = 5;        // ç¿»è¨³å…ˆè¨€èª
  string engine_name = 6;              // ä½¿ç”¨ã•ã‚ŒãŸç¿»è¨³ã‚¨ãƒ³ã‚¸ãƒ³å
  float confidence_score = 7;          // ç¿»è¨³ã®ä¿¡é ¼åº¦ã‚¹ã‚³ã‚¢ (-1.0 = æœªè¨­å®š)
  int64 processing_time_ms = 8;        // ç¿»è¨³å‡¦ç†æ™‚é–“ï¼ˆãƒŸãƒªç§’ï¼‰
  bool is_success = 9;                 // ç¿»è¨³ãŒæˆåŠŸã—ãŸã‹ã©ã†ã‹
  TranslationError error = 10;         // ã‚¨ãƒ©ãƒ¼æƒ…å ±ï¼ˆoptionalï¼‰
  map<string, string> metadata = 11;   // ãƒ¡ã‚¿ãƒ‡ãƒ¼ã‚¿ï¼ˆJSONæ–‡å­—åˆ—åŒ–ï¼‰
  google.protobuf.Timestamp timestamp = 12; // ãƒ¬ã‚¹ãƒãƒ³ã‚¹ã‚¿ã‚¤ãƒ ã‚¹ã‚¿ãƒ³ãƒ—
}

// ============================================================================
// Request/Response Messages - Batch Translation
// ============================================================================

/// <summary>
/// ãƒãƒƒãƒç¿»è¨³ãƒªã‚¯ã‚¨ã‚¹ãƒˆ
/// </summary>
message BatchTranslateRequest {
  repeated TranslateRequest requests = 1; // ç¿»è¨³ãƒªã‚¯ã‚¨ã‚¹ãƒˆã®ãƒªã‚¹ãƒˆ
  string batch_id = 2;                    // ãƒãƒƒãƒID (UUIDæ–‡å­—åˆ—)
  google.protobuf.Timestamp timestamp = 3; // ãƒªã‚¯ã‚¨ã‚¹ãƒˆã‚¿ã‚¤ãƒ ã‚¹ã‚¿ãƒ³ãƒ—
}

/// <summary>
/// ãƒãƒƒãƒç¿»è¨³ãƒ¬ã‚¹ãƒãƒ³ã‚¹
/// </summary>
message BatchTranslateResponse {
  repeated TranslateResponse responses = 1; // ç¿»è¨³ãƒ¬ã‚¹ãƒãƒ³ã‚¹ã®ãƒªã‚¹ãƒˆ
  string batch_id = 2;                      // å¯¾å¿œã™ã‚‹ãƒãƒƒãƒID
  int32 success_count = 3;                  // æˆåŠŸã—ãŸç¿»è¨³ã®æ•°
  int32 failure_count = 4;                  // å¤±æ•—ã—ãŸç¿»è¨³ã®æ•°
  int64 total_processing_time_ms = 5;       // ç·å‡¦ç†æ™‚é–“ï¼ˆãƒŸãƒªç§’ï¼‰
  google.protobuf.Timestamp timestamp = 6;  // ãƒ¬ã‚¹ãƒãƒ³ã‚¹ã‚¿ã‚¤ãƒ ã‚¹ã‚¿ãƒ³ãƒ—
}

// ============================================================================
// Request/Response Messages - Health Check
// ============================================================================

/// <summary>
/// ãƒ˜ãƒ«ã‚¹ãƒã‚§ãƒƒã‚¯ãƒªã‚¯ã‚¨ã‚¹ãƒˆ
/// </summary>
message HealthCheckRequest {
  // Empty message - health check doesn't need parameters
}

/// <summary>
/// ãƒ˜ãƒ«ã‚¹ãƒã‚§ãƒƒã‚¯ãƒ¬ã‚¹ãƒãƒ³ã‚¹
/// </summary>
message HealthCheckResponse {
  bool is_healthy = 1;                      // ã‚µãƒ¼ãƒãƒ¼ãŒæ­£å¸¸ã‹ã©ã†ã‹
  string status = 2;                        // ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸
  map<string, string> details = 3;          // è©³ç´°æƒ…å ±
  google.protobuf.Timestamp timestamp = 4;  // ã‚¿ã‚¤ãƒ ã‚¹ã‚¿ãƒ³ãƒ—
}

/// <summary>
/// æº–å‚™çŠ¶æ…‹ç¢ºèªãƒªã‚¯ã‚¨ã‚¹ãƒˆ
/// </summary>
message IsReadyRequest {
  // Empty message - ready check doesn't need parameters
}

/// <summary>
/// æº–å‚™çŠ¶æ…‹ç¢ºèªãƒ¬ã‚¹ãƒãƒ³ã‚¹
/// </summary>
message IsReadyResponse {
  bool is_ready = 1;                        // ã‚µãƒ¼ãƒãƒ¼ãŒæº–å‚™å®Œäº†ã‹ã©ã†ã‹
  string status = 2;                        // ã‚¹ãƒ†ãƒ¼ã‚¿ã‚¹ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸
  map<string, string> details = 3;          // è©³ç´°æƒ…å ±ï¼ˆãƒ¢ãƒ‡ãƒ«ãƒ­ãƒ¼ãƒ‰çŠ¶æ³ãªã©ï¼‰
  google.protobuf.Timestamp timestamp = 4;  // ã‚¿ã‚¤ãƒ ã‚¹ã‚¿ãƒ³ãƒ—
}

// ============================================================================
// Service Definition
// ============================================================================

/// <summary>
/// Baketaç¿»è¨³ã‚µãƒ¼ãƒ“ã‚¹
/// Phase 2: gRPCé€šä¿¡ã«ã‚ˆã‚‹Pythonç¿»è¨³ã‚µãƒ¼ãƒãƒ¼ã¨ã®é€£æº
/// </summary>
service TranslationService {
  /// <summary>
  /// å˜ä¸€ãƒ†ã‚­ã‚¹ãƒˆã‚’ç¿»è¨³ã—ã¾ã™
  /// </summary>
  rpc Translate(TranslateRequest) returns (TranslateResponse);

  /// <summary>
  /// è¤‡æ•°ã®ãƒ†ã‚­ã‚¹ãƒˆã‚’ãƒãƒƒãƒç¿»è¨³ã—ã¾ã™
  /// </summary>
  rpc TranslateBatch(BatchTranslateRequest) returns (BatchTranslateResponse);

  /// <summary>
  /// ã‚µãƒ¼ãƒãƒ¼ã®ãƒ˜ãƒ«ã‚¹ãƒã‚§ãƒƒã‚¯ã‚’å®Ÿè¡Œã—ã¾ã™
  /// </summary>
  rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);

  /// <summary>
  /// ã‚µãƒ¼ãƒãƒ¼ãŒæº–å‚™å®Œäº†ã‹ã©ã†ã‹ã‚’ç¢ºèªã—ã¾ã™
  /// </summary>
  rpc IsReady(IsReadyRequest) returns (IsReadyResponse);

  // ============================================================================
  // ğŸ”¥ [FUTURE] Phase 12.2: ã‚¹ãƒˆãƒªãƒ¼ãƒŸãƒ³ã‚°ç¿»è¨³ã‚µãƒãƒ¼ãƒˆï¼ˆäºˆç´„ï¼‰
  // ============================================================================
  // /// <summary>
  // /// ã‚¹ãƒˆãƒªãƒ¼ãƒŸãƒ³ã‚°ç¿»è¨³ã‚’å®Ÿè¡Œã—ã¾ã™ï¼ˆåŒæ–¹å‘ã‚¹ãƒˆãƒªãƒ¼ãƒŸãƒ³ã‚°ï¼‰
  // /// </summary>
  // rpc TranslateStream(stream TranslateRequest) returns (stream TranslateResponse);
}
